docs/new_flux_ideas_and_guide.md

```diff
@@ -0,0 +1,985 @@
+| CA | 2020-02-22T17:00
+| CA | 2020-02-22T18:00
+| WA | 2020-02-22T18:00
+
+We can see that CA reported in all three hours while both CA and WA reported in only the last hour. Also see that 
+distinct discards columns that aren't in the `by` argument or the `distinct` column itself. So we use the `time` 
+name for where we put the window markers for each distinct state. You could then count the number of states reporting 
+per hour:
+
+```
+@my_data.h2o_temperature{state, time > -3h}
+    |> select(fn: distinct(state), window: 1h)
+    |> timeShift(-1m)
+    |> count("state", window: 1h)
+```
+
+Would produce:
+
+| state | time
+| --- | --- |
+| 1 | 2020-02-22T16:00
+| 1 | 2020-02-22T17:00
+| 2 | 2020-02-22T18:00
+
+See that we had to use the `timeShift` function to move the timestamps of the windows back so that the resulting 
+times would be accurate for `aggregate`.
+
+All of this is encapsulated by the `countDistinct` function. For example, if we wanted to count the number of 
+distinct locations that reported in each sate for each hour in the last three hours.
+
+```
+@my_data.h2o_temperature{location, state, bottom_degrees, time > -3h}
+    |> countDistinct(location, by: ["state"], window: 1h)
+```
+
+### Math, basic transformations, and using previous results in new queries
+Flux makes it easy to do basic math and transformations across tables of data. First, let's look at 
+renaming columns. All of the different predicate blocks recognize the `as` keyword to rename. For 
+example:
+
+```
+@my_data.h2o_temperature{location, state, bottom_degrees as degrees, top_degrees, time}
+
+// or renaming columns in an aggregate:
+@my_data.h2o_temperature{bottom_degrees, time > -1h}
+    |> aggregate({min(bottom_degrees) as min, max(bottom_degrees) as max})
+```
+
+Here are some examples of doing math and introducing the concept of variables:
+
+```
+temps = @my_data.h2o_temperature{bottom_degrees, surface_degrees, time, location == "coyote_creek"}
+
+// do some math and return a new table with an addition column. Note that as is required on the end. Resulting
+// table will have all the columns of the previous on and this new one
+new_table = temps.surface_degrees - temps.bottom_degrees as bottom_surface_difference
+
+// or you can pipe forward it to another function
+temps.surface_degrees - temps.bottom_degrees as bottom_surface_difference
+    |> sort(by: ["bottom_surface_differrence"])
+
+// convert from fahrenheit to celsius and replace the value previously held by bottom_degrees
+(temps.bottom_degrees - 32) * 5 / 9 as bottom_degrees
+```
+
+The above example shows assigning a result set to a variable so that it can be used multiple times later. You 
+can also use this construction to do implicit joins across tables. Here's a made up example:
+
+```
+foo_data = @my_data.foo{value as foo_val, and time > -1h}
+bar_data = @my_data.bar{value as bar_val, and time > -1h}
+
+foo_data.foo_val - bar_data.bar_val as difference_to_bar
```

Comment:
this is one issue I have had with flux, its implicit return/yield syntax. This seems to make this even more complicated to me 🤔. 

Feels very magical/unobvious that this adds to the returned group imo 🤷‍♂

Reply:
How would you like to see it done? Are you wanting an explicit `return` statement or something? 

I hadn't heard this feedback before and I quite like the implicit nature as it makes writing simple queries easier, but I want to see it from your perspective if I can. Any suggestions on how this could be more explicit?

Reply:
I'm not sure off hand, have to noodle on it a bit. The current state of things I can wrap my head around nowadays.

However, this feels very strange to me:

```
foo_data.foo_val - bar_data.bar_val as difference_to_bar
```

in my head I'd expect to only get `difference_to_bar` here and not the result shared in the doc.

```
| foo_val | bar_val | difference_to_bar | time
| --- | --- | --- | --- | 
```

does that make sense why I find it confusing? 

This is one thing I liked about the map func, it has the `with` syntax, which makes things very obvious when reading it that you'll get everything and whatever is extending it.  Perhaps something like that could be used here? I don't know specifically what that may look like though 🤔

Reply:
I agree with @jsteenb2 here that I would expect to only get `difference_to_bar` back here.